include/seqan3/core/metafunction/function.hpp

```diff
@@ -0,0 +1,70 @@
+//                 SeqAn - The Library for Sequence Analysis
+// ============================================================================
+//
+// Copyright (c) 2006-2018, Knut Reinert & Freie Universitaet Berlin
+// Copyright (c) 2016-2018, Knut Reinert & MPI Molekulare Genetik
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//     * Neither the name of Knut Reinert or the FU Berlin nor the names of
+//       its contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL KNUT REINERT OR THE FU BERLIN BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ============================================================================
+
+/*!\file
+ * \brief Provides various metafunctions for use on functions.
+ * \author Hannes Hauswedell <hannes.hauswedell AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <seqan3/core/platform.hpp>
+
+// ----------------------------------------------------------------------------
+// is_constexpr
+// ----------------------------------------------------------------------------
+
+namespace seqan3::detail
+{
+
+//!\brief NO-OP helper for is_constexpr
+//!\ingroup metafunction
+constexpr auto is_constexpr_helper(int) {}
+
+} // namespace seqan3::detail
+
+namespace seqan3
+{
+
+/*!\brief Returns true if the expression passed to this macro can be evaluated at compile time, false otherwise.
+ * \ingroup metafunction
+ * \returns true or false.
+ */
+#if defined(__clang__) // bug: https://bugs.llvm.org//show_bug.cgi?id=15481
+#define is_constexpr(...) __builtin_constant_p( (__VA_ARGS__,0) )
+#else
+#define is_constexpr(...) noexcept(seqan3::detail::is_constexpr_helper( (__VA_ARGS__, 0) ))
```

Comment:
Mhmm, https://en.cppreference.com/w/cpp/language/constexpr suggests, that this is possible util c++17?
And the same states documentation for the https://en.cppreference.com/w/cpp/language/noexcept operator.
So as far as I understand, you simply check if any of the expressions added to __VA_ARGS__ can be evaluated at compile time and if so the noexcept returns true. But now I am irritated by the limitations of cppref, also because there wasn't any `since` clause what to expect since c++17.
Maybe you can elaborate?